browser/ui/webui/brave_new_tab/new_tab_ui.cc

```diff
@@ -0,0 +1,110 @@
+#include "brave/components/brave_new_tab/resources/grit/brave_new_tab_generated_map.h"
+#include "brave/components/l10n/common/localization_util.h"
+#include "brave/components/ntp_background_images/browser/ntp_custom_images_source.h"
+#include "chrome/browser/profiles/profile.h"
+#include "chrome/browser/ui/webui/webui_util.h"
+#include "components/grit/brave_components_resources.h"
+#include "components/grit/brave_components_strings.h"
+#include "components/strings/grit/components_strings.h"
+#include "content/public/browser/navigation_entry.h"
+#include "content/public/browser/web_contents.h"
+#include "content/public/browser/web_ui.h"
+#include "content/public/browser/web_ui_data_source.h"
+#include "ui/base/webui/web_ui_util.h"
+
+namespace brave_new_tab {
+
+namespace {
+
+static constexpr webui::LocalizedString kStrings[] = {
+    {"backgroundSettingsTitle", IDS_NEW_TAB_BACKGROUND_SETTINGS_TITLE},
+    {"braveBackgroundLabel", IDS_NEW_TAB_BRAVE_BACKGROUND_LABEL},
+    {"customBackgroundLabel", IDS_NEW_TAB_CUSTOM_BACKGROUND_LABEL},
+    {"customBackgroundTitle", IDS_NEW_TAB_CUSTOM_BACKGROUND_LABEL},
+    {"gradientBackgroundLabel", IDS_NEW_TAB_GRADIENT_BACKGROUND_LABEL},
+    {"gradientBackgroundTitle", IDS_NEW_TAB_GRADIENT_BACKGROUND_LABEL},
+    {"photoCreditsText", IDS_NEW_TAB_PHOTO_CREDITS_TEXT},
+    {"randomizeBackgroundLabel", IDS_NEW_TAB_RANDOMIZE_BACKGROUND_LABEL},
+    {"settingsTitle", IDS_NEW_TAB_SETTINGS_TITLE},
+    {"showBackgroundsLabel", IDS_NEW_TAB_SHOW_BACKGROUNDS_LABEL},
+    {"showSponsoredImagesLabel", IDS_NEW_TAB_SHOW_SPONSORED_IMAGES_LABEL},
+    {"solidBackgroundLabel", IDS_NEW_TAB_SOLID_BACKGROUND_LABEL},
+    {"solidBackgroundTitle", IDS_NEW_TAB_SOLID_BACKGROUND_LABEL},
+    {"uploadBackgroundLabel", IDS_NEW_TAB_UPLOAD_BACKGROUND_LABEL},
+};
+
+// Adds support for displaying images stored in the custom background image
+// folder.
+void AddCustomImageDataSource(Profile* profile) {
+  auto* custom_background_service =
+      BraveNTPCustomBackgroundServiceFactory::GetForContext(profile);
+  if (!custom_background_service) {
+    return;
+  }
+  auto source = std::make_unique<ntp_background_images::NTPCustomImagesSource>(
+      custom_background_service);
+  content::URLDataSource::Add(profile, std::move(source));
+}
+
+}  // namespace
+
+NewTabUI::NewTabUI(content::WebUI* web_ui) : ui::MojoWebUIController(web_ui) {
+  auto* profile = Profile::FromWebUI(web_ui);
+
+  auto* source = content::WebUIDataSource::CreateAndAdd(
+      Profile::FromWebUI(web_ui), chrome::kChromeUINewTabHost);
+
+  webui::SetupWebUIDataSource(source, kBraveNewTabGenerated,
+                              IDR_BRAVE_NEW_TAB_HTML);
+
+  source->OverrideContentSecurityPolicy(
+      network::mojom::CSPDirectiveName::ImgSrc,
+      "img-src chrome://resources chrome://theme chrome://background-wallpaper "
+      "chrome://custom-wallpaper chrome://branded-wallpaper chrome://favicon2 "
+      "blob: data: 'self';");
```

Comment:
what's the reason for allowing a fully unrestricted `blob:` url here? The `data:` makes sense so that we can embed the images here in the HTML, but I wasn't following the purpose of `blob:` usage here.

Reply:
Thanks for checking on this. `blob:` URLs are used by both Brave News and the search box to display images coming from the private CDN.

Reply:
Gotcha ok, thanks for the additional info. This CSP seems good to me.